src/rand.rs

```diff
@@ -83,21 +84,64 @@ pub(crate) mod sealed {
         fn as_mut_bytes(&mut self) -> &mut [u8]; // `AsMut<[u8]>::as_mut`
     }
 
-    macro_rules! impl_random_arrays {
-        [ $($len:expr)+ ] => {
-            $(
-                impl RandomlyConstructable for [u8; $len] {
-                    #[inline]
-                    fn zero() -> Self { [0; $len] }
+    // Safety: in order for a type to be POD, it must satisfy the following criteria:
+    // - The type must be inhabited (ie not the never type)
+    // - All bit patterns must be valid
+    // - The type must have no padding
+    // - Any fields within the type must also be Pod
+    pub unsafe trait Pod: Sized {
+        fn zero() -> Self;
+    }
 
+    macro_rules! impl_pod {
+        ($($type:ty),*) => {
+            $(
+                unsafe impl Pod for $type {
                     #[inline]
-                    fn as_mut_bytes(&mut self) -> &mut [u8] { &mut self[..] }
+                    fn zero() -> Self { 0 }
                 }
-            )+
-        }
+            )*
+        };
     }
 
-    impl_random_arrays![4 8 16 32 48 64 128 256];
+    impl_pod!(i8, u8, i16, u16, i32, u32, i64, u64, i128, u128, isize, usize);
+
+    impl<T: Pod + Copy, const N: usize> RandomlyConstructable for [T; N] {
+        #[inline]
+        fn zero() -> Self {
+            // Note that the Copy bound on T is required in order to initialize this array
+            // without unsafe code.
+            [T::zero(); N]
+        }
+
+        #[inline]
+        fn as_mut_bytes(&mut self) -> &mut [u8] {
+            // Since all values in this expression are constant, the compiler should optimize this
+            // assertion out.
+            assert!(
+                N == 0 || (isize::MAX as usize) / N >= mem::size_of::<T>(),
+                "Array too large to be converted to slice of u8"
+            );
```

Comment:
Is this assertion necessary? I thought it was already UB to have `self` be larger than isize bytes?

Reply:
We should have tests for the case where N == 0 and where `core::mem::size_of<T>` is zero.

Reply:
When I want to be sure the compiler can evaluate an expression at compile time, I do:
```
const NO_OVERFLOW: bool = N == 0 || (isize::MAX as usize) / N >= mem::size_of::<T>();
assert!(NO_OVERFLOW);
```
I'd rather not have the string in the assertion. I'm working to eliminate all such strings.

Reply:
> Is this assertion necessary? I thought it was already UB to have `self` be larger than isize bytes?

Ope I somehow managed to not read the `* mem::size_of<T>` part in the docs. I'll take that out - it makes matters simpler anyway.